src/auth/ha-authorize.ts

```diff
@@ -1,251 +1,516 @@
     }
   }
 
   protected render() {
-    if (this._error) {
+    if (this._error)
       return html`
-        <style>
-          ha-authorize ha-alert {
-            display: block;
-            margin: 16px 0;
-          }
-        </style>
-        <ha-alert alert-type="error"
-          >${this._error} ${this.redirectUri}</ha-alert
-        >
+        ${makeIcon(mdiAlertCircleOutline, "error")}
+        <p>${this._error}</p>
       `;
-    }
-
-    if (!this._authProviders) {
+    if (this._finished)
       return html`
-        <style>
-          ha-authorize p {
-            font-size: 14px;
-            line-height: 20px;
-          }
-        </style>
-        <p>${this.localize("ui.panel.page-authorize.initializing")}</p>
+        ${makeIcon(mdiCheck, "info")}
+        <p>${this.localize("ui.panel.page-authorize.logging_in")}</p>
       `;
-    }
 
-    const inactiveProviders = this._authProviders.filter(
-      (prv) => prv !== this._authProvider
+    if (!this._authProviders || !this._authProvider) return nothing;
+
+    const { redirectUri, clientId, mode } = initialState as InitialState;
+    const header = this.localize(
+      `ui.panel.page-authorize.heading_${mode.name}`
     );
 
-    const app = this.clientId && this.clientId in appNames;
+    if (!header) return nothing;
 
-    return html`
-      <style>
-        ha-pick-auth-provider {
-          display: block;
-          margin-top: 48px;
-        }
-        ha-auth-flow {
-          display: block;
-          margin-top: 24px;
-        }
-        ha-alert {
-          display: block;
-          margin: 16px 0;
+    let error: HTMLTemplateResult | string | undefined;
+    const fields: { item: HaFormSchema; name: string }[] = [];
+    if (this._step?.type === "form") {
+      if (this._step.errors?.base) {
+        error = this.localize(
+          `ui.panel.page-authorize.form.providers.${this._authProvider.type}.error.${this._step.errors.base}`
+        );
+      }
+      for (const item of this._step.data_schema) {
+        const name = this.localize(
+          `ui.panel.page-authorize.form.providers.${this._authProvider.type}.step.${this._step.step_id}.data.${item.name}`
+        );
+        fields.push({ item, name });
+      }
+    } else if (this._step?.type === "abort") {
+      error = html`${this.localize("ui.panel.page-authorize.abort_intro")}
+        <br />
+        ${this.localize(
+          `ui.panel.page-authorize.form.providers.${this._authProvider.type}.abort.${this._step.reason}`
+        )}`;
+    }
+
+    const buttons: HTMLTemplateResult[] = [];
+    for (const provider of this._authProviders) {
+      if (provider === this._authProvider) continue;
+      // eslint-disable-next-line @typescript-eslint/no-loop-func
+      const useOtherMethod = async () => {
+        const result = await loadStep(
+          this._step,
+          clientId,
+          provider,
+          redirectUri
+        );
```

Comment:
We don't use inline event handlers, this should be a class function

Reply:
It's not that simple. Each button performs a different function. While I could do something like moving the core to the class and making a wrapper function, binding the function, or identifying the auth provider from the element's text, I think it's the most reasonable to do it this way.